tests/ci_build/lint_cpp.py

```diff
@@ -0,0 +1,173 @@
+        _ = cpplint.ParseArguments(cpplint_args)
+        cpplint._SetFilters(
+            ",".join(
+                [
+                    "-build/c++11",
+                    "-build/include,",
+                    "+build/namespaces",
+                    "+build/include_what_you_use",
+                    "+build/include_order",
+                ]
+            )
+        )
+        cpplint._SetCountingStyle("toplevel")
+        cpplint._line_length = 100
+
+    def process_cpp(self, path, suffix):
+        """Process a cpp file."""
+        _cpplint_state.ResetErrorCounts()
+        cpplint.ProcessFile(str(path), _cpplint_state.verbose_level)
+        _cpplint_state.PrintErrorCounts()
+        errors = _cpplint_state.errors_by_category.copy()
+
+        if suffix == "h":
+            self.cpp_header_map[str(path)] = errors
+        else:
+            self.cpp_src_map[str(path)] = errors
+
+    @staticmethod
+    def _print_summary_map(strm, result_map, ftype):
+        """Print summary of certain result map."""
+        if len(result_map) == 0:
+            return 0
+        npass = sum(1 for x in result_map.values() if len(x) == 0)
+        strm.write(f"====={npass}/{len(result_map)} {ftype} files passed check=====\n")
+        for fname, emap in result_map.items():
+            if len(emap) == 0:
+                continue
+            strm.write(
+                f"{fname}: {sum(emap.values())} Errors of {len(emap)} Categories map={str(emap)}\n"
+            )
+        return len(result_map) - npass
+
+    def print_summary(self, strm):
+        """Print summary of lint."""
+        nerr = 0
+        nerr += Lint._print_summary_map(strm, self.cpp_header_map, "cpp-header")
+        nerr += Lint._print_summary_map(strm, self.cpp_src_map, "cpp-source")
+        if nerr == 0:
+            strm.write("All passed!\n")
+        else:
+            strm.write(f"{nerr} files failed lint\n")
+        return nerr
+
+
+_HELPER = Lint()
+
+cpplint.GetHeaderGuardCPPVariable = get_header_guard_dmlc
+
+
+def process(fname, allow_type):
+    """Process a file."""
+    fname = str(fname)
+    arr = fname.rsplit(".", 1)
+    if fname.find("#") != -1 or arr[-1] not in allow_type:
+        return
+    if arr[-1] in CXX_SUFFIX:
+        _HELPER.process_cpp(fname, arr[-1])
+
+
+def main():
+    parser = argparse.ArgumentParser(description="run cpp lint")
+    parser.add_argument(
+        "path",
+        nargs="*",
+        help="Path to traverse",
+        default=["src", "include", os.path.join("R-package", "src"), "python-package", "plugin/sycl"],
+    )
+    parser.add_argument(
+        "--exclude_path",
+        nargs="+",
+        default=[],
+        help="exclude this path, and all subfolders if path is a folder",
+    )
+    args = parser.parse_args()
+    excluded_paths = filepath_enumerate(args.exclude_path)
+
+    allow_type = []
+    allow_type += CXX_SUFFIX
+
+    for path in args.path:
```

Comment:
I would suggest to consistently convert path strings into `pathlib.Path` objects. They come with a few benefits as pointed out in other comments.